### ISSUE TITLE
Invalid Unicode path encountered

### ISSUE THREAD

**Issue by lasselammi** [GoogleCloudPlatform/gsutil#221]
I'm trying to upload a file with filename "kuormatraktorityön suunnittelun työmalli.pdf" and I get the following error message: "CommandException: Invalid Unicode path encountered". I'm using the most recent version of gsutil. I was able to upload this same file with a previous version of gsutil. Why is this a problem now?

**Comment by SergeAx** [comment]
@jterrace - I've got the same issue on my Windows machine while trying to use gsutil rsync as a cloud backup solution:

```
CommandException: Invalid Unicode path encountered
('Z:\\Backup\\neo.sergeax.ru\\IPub\\a.gu.ru\\media\\D__\xcb\xe5\xe2\xe8\xed_rcrc.jpg').
```

The real file name is "D__Левин_rcrc.jpg" (it's Russan) Here's the output of getdefaultlocale:

```
C:\gsutil>python -c "import locale; print locale.getdefaultlocale()"
('ru_RU', 'cp1251')
```

What can I do to rsync this file (and thousands like it) to Google Cloud Storage? Thank you!

**Comment by jterrace** [comment]
cp1251 is not a supported byte encoding. You'll need to convert your filenames to UTF-8 before uploading.

**Comment by SergeAx** [comment]
Is that possible at all on NTFS filesystem?

**Comment by jterrace** [comment]
Yes, NTFS supports unicode. You could try the Windows version of the iconv program:
http://gnuwin32.sourceforge.net/packages/libiconv.htm

**Comment by SergeAx** [comment]
Sorry, it's too unobvious for me. I believe it was Plato who said that to google a question one should know half of the answer :) After some research online I still know about 10%, clearly not enough. Can you point me to manual or article for backing up my non-UTF named files to Google Cloud Storage using gsutil in rsync mode?

Thank you very much anyway.

**Comment by moander** [comment]
I have the same problem on Windows, but if I mount the same disk on osx using smb it works just fine.

**Comment by eblazer** [comment]
I too experience this error, but ONLY when executed from crontab. When I manually run the same script from my shell, I don't get this error.

"Starting Google Cloud Storage RSync
Mon Dec  7 10:00:01 PST 2015
Building synchronization state...
Caught non-retryable exception while listing file:///volumeUSB1/usbshare/Test_1: 'ascii' codec can't encode character u'\u2019' in position 81: ordinal not in range(128)
At destination listing 10000...
"

**Comment by eblazer** [comment]
Discovered the issue, per `gsutil help encoding`

  Unicode errors for valid Unicode filepaths can be caused by lack of Python
  locale configuration on Linux and Mac OSes. If your file paths are Unicode
  and you get encoding errors, **ensure the LANG environment variable is set
  correctly. Typically, the LANG variable should be set to something like
  "en_US.UTF-8" or "de_DE.UTF-8".**

My script being called in the crontab context did not have a LANG env var.

**Comment by fredrikaverpil** [comment]
Wow, how odd. I'm running `gsutil rsync` from within a `centos:7` Docker container.

For some reason, `LANG` is not set and I get this:

```python
>>> python -c "import locale; print locale.getdefaultlocale()"
(None, None)
```

Thanks to @eblazer this is now solved, by setting `LANG=en_US.UTF-8` 👍
